File: Leetcode/cpp/medium/3296.cpp
/*
 * @lc app=leetcode.cn id=3296 lang=cpp
 *
 * [3296] 移山所需的最少秒数
 */


// @lcpr-template-start
#include <bits/stdc++.h>
using namespace std;
// @lcpr-template-end
// @lc code=start
class Solution1 {
public:
    long long minNumberOfSeconds(int mountainHeight, vector<int>& workerTimes) {
        int n = workerTimes.size();
        auto check = [&](long long m) -> bool {
            long long tot = 0;
            for (int t : workerTimes) {
                tot += (-1 + sqrt(1 + 8 * m / t)) / 2;
                if (tot >= mountainHeight) return true;
            }
            return false;
        };

        long long h = ceil(double(mountainHeight) / n);
        long long left = 0;
        long long right = ranges::max(workerTimes) * h * (h + 1) / 2;
        while (left <= right) {
            long long mid = left + (right - left) / 2;
            if (check(mid)) right = mid - 1;
            else left = mid + 1;
        }
        return left;
    }
};

class Solution2 {
public:
    long long minNumberOfSeconds(int mountainHeight, vector<int>& workerTimes) {
        priority_queue<array<long long, 3>, vector<array<long long, 3>>, greater<>> pq;
        for (int t : workerTimes) {
            pq.emplace(array<long long, 3>{t, t, 1});
        }

        long long ans = 0;
        while (mountainHeight--) {
            auto [t, d, m] = pq.top(); pq.pop();
            m += 1;
            ans = t;
            pq.push({t + d * m, d, m});
        }
        return ans;
    }
};

// using Solution = Solution1;
using Solution = Solution2;
// @lc code=end

File: Leetcode/rust/medium/3296.rs
/*
 * @lc app=leetcode id=3296 lang=rust
 *
 * [3296] Minimum Number of Seconds to Make Mountain Height Zero
 */


// @lcpr-template-start
struct Solution;
// @lcpr-template-end
// @lc code=start
struct Solution1;
impl Solution1 {
    pub fn min_number_of_seconds(mountain_height: i32, worker_times: Vec<i32>) -> i64 {
        let mountain_height = mountain_height as i64;
        let check = |m: i64| -> bool {
            let mut tot: i64 = 0;
            for &t in &worker_times {
                tot += ((-1 + (1 + 8 * m / (t as i64)).isqrt()) / 2) as i64;
                if tot >= mountain_height {
                    return true;
                }
            }
            false
        };

        let h: i64 = (mountain_height as f64 / worker_times.len() as f64).ceil() as i64;
        let max_t: i64 = *worker_times.iter().max().unwrap() as i64;

        let mut left: i64 = 0;
        let mut right: i64 = max_t * h * (h + 1) / 2;
        while left <= right {
            let mid = left + (right - left) / 2;
            if check(mid) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        left
    }
}

use std::collections::BinaryHeap;
use std::cmp::Reverse;

struct Solution2;
impl Solution2 {
    pub fn min_number_of_seconds(mountain_height: i32, worker_times: Vec<i32>) -> i64 {
        let mut hp = BinaryHeap::new();
        for &t in &worker_times {
            hp.push(Reverse((t as i64, t as i64, 1)));
        }

        let mut ans: i64 = 0;
        for _ in 0..mountain_height {
            let Reverse((t, d, m)) = hp.pop().unwrap();
            ans = t;
            hp.push(Reverse((t + d * (m + 1), d, m + 1)));
        }
        ans
    }
}

impl Solution {
    pub fn min_number_of_seconds(mountain_height: i32, worker_times: Vec<i32>) -> i64 {
        Solution1::min_number_of_seconds(mountain_height, worker_times)
        // Solution2::min_number_of_seconds(mountain_height, worker_times)
    }
}
// @lc code=end

